improver/calibration/reliability_calibration.py

```diff
@@ -0,0 +1,414 @@
+#   contributors may be used to endorse or promote products derived from
+#   this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Reliability calibration plugins."""
+
+import iris
+from iris.exceptions import CoordinateNotFoundError
+import numpy as np
+
+from improver import BasePlugin
+from improver.utilities.cube_manipulation import MergeCubes
+from improver.metadata.utilities import generate_mandatory_attributes
+from improver.metadata.probabilistic import find_threshold_coordinate
+from improver.calibration.utilities import filter_non_matching_cubes
+
+
+class ConstructReliabilityCalibrationTables(BasePlugin):
+
+    """A plugin for creating and populating reliability calibration tables."""
+
+    def __init__(self, n_probability_bins=5, single_value_limits=True):
+        """
+        Initialise class for creating reliability calibration tables. These
+        tables include data columns entitled observation_count,
+        sum_of_forecast_probabilities, and forecast_count, defined below.
+
+        n_probability_bins (int):
+            The total number of probability bins required in the reliability
+            tables. If single value limits are turned on, these are included in
+            this total.
+        single_value_limits (bool):
+            Mandates that the extrema bins (0 and 1) should be single valued,
+            with a small precision tolerance, defined below as
+            single_value_tolerance. This gives bins of 0 to
+            single_value_tolerance and (1 - single-value_tolerance) to 1.
+        """
+        self.single_value_tolerance = 1.0E-6
+        self.probability_bins = self._define_probability_bins(
+            n_probability_bins, single_value_limits)
+        self.table_columns = np.array(
+            ['observation_count', 'sum_of_forecast_probabilities',
+             'forecast_count'])
+        self.expected_table_shape = (len(self.table_columns),
+                                     n_probability_bins)
+
+    def __repr__(self):
+        """Represent the configured plugin instance as a string."""
+        bin_values = ', '.join(
+            ['[{:1.2f} --> {:1.2f}]'.format(*item)
```

Comment:
Somewhere we need to state in a docstring which way round the bounds are inclusive. (a<=prob< b or a<prob<=b) although it will make very little difference in practise.

Reply:
You're right, it won't make much difference. I was confused earlier when talking to you, as I was referring to the observation thresholding rather than the comparison with probability bin boundaries. The equality operator used is actually always bounds inclusive at both ends as the requirement of the DimCoord to have non-overlapping bins means that there is no overlap.

What this actually translates to is rather hard to say concisely. It must be "greater than" for the first non-zero bin, else nothing could fall in the zero bin (though it does in fact have a finite width). It must be "less than" at the top end, otherwise nothing could fall in the 100% bin. Hence my fudge of showing the ranges and being a bit quiet beyond that.

Any advice?

Reply:
Agreed to put a little bit more info in the extended documentation